NLog.Windows.Forms/RichTextBoxTarget.cs

```diff
@@ -299,6 +318,100 @@ public RichTextBoxTargetMessageRetentionStrategy MessageRetention
+        /// <seealso cref="LinkClicked"/>
+        /// </summary>
+        [DefaultValue(false)]
+        public bool SupportLinks 
+        {
+            get { return supportLinks; }
+            set
+            {
+                supportLinks = value;
+                if (supportLinks)
+                {
+                    lock (linkedEventsLock)
+                    {
+                        if (linkedEvents == null)
+                        {
+                            linkedEvents = new Dictionary<int, LogEventInfo>();
+                        }
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Type of delegate for <see cref="LinkClicked"/> event.
+        /// </summary>
+        /// <param name="sender">The target that caused the event</param>
+        /// <param name="linkText">Visible text of the link being clicked</param>
+        /// <param name="logEvent">Original log event that caused a line with the link</param>
+        public delegate void DelLinkClicked(RichTextBoxTarget sender, string linkText, LogEventInfo logEvent);
+
+        /// <summary>
+        /// Event fired when the user clicks on a link in the control created by the "rtb-link" renderer (<see cref="RichTextBoxLinkLayoutRenderer"/>).
+        /// <seealso cref="DelLinkClicked"/>
+        /// </summary>
+        public event DelLinkClicked LinkClicked;
+
+        /// <summary>
+        /// Actual value of the <see cref="LinkClicked"/> property
+        /// </summary>
+        private bool supportLinks = false;
+
+        /// <summary>
+        /// Lock for <see cref="linkedEvents"/> dictionary access
+        /// </summary>
+        private readonly object linkedEventsLock = new object();
+
+        /// <summary>
+        /// A map from link id to a corresponding log event
+        /// </summary>
+        private Dictionary<int, LogEventInfo> linkedEvents;
+
+        /// <summary>
+        /// Returns number of events stored for active links in the control. 
+        /// Used only in tests to check that non needed events are removed.
+        /// </summary>
+        internal int? LinkedEventsCount
+        {
+            get
+            {
+                lock (linkedEventsLock)
+                {
+                    if (linkedEvents == null)
+                    {
+                        return null;
+                    }
+                    return linkedEvents.Count;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Internal prefix that is added to the link id in RTF
+        /// </summary>
+        private const string LINK_PREFIX = "link";
+
+
+        /// <summary>
+        /// Used to capture link placeholders in <see cref="SendTheMessageToRichTextBox"/>
+        /// </summary>
+        private readonly Regex linkAddRegex = new Regex(@"(\([a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}\))", RegexOptions.Compiled);
```

Comment:
I was not aware about caching functional, so that you for pointing this out. I've read (Compilation and Reuse in Regular Expressions)[https://msdn.microsoft.com/en-us/library/8zbs0h2f(v=vs.110).aspx] and (Best Practices for Regular Expressions in the .NET Framework)[https://msdn.microsoft.com/en-us/library/gg578045(v=vs.110).aspx] and I still don't think there's a reason to not use compiled regexes in this case.

if I get the text right, the regex cache is just 15 items. So it looks like a valuable user-space resource. And consuming 20% of it for some helper library code looks like a waste. I think well-behaving libraries should not do such things.

I think that these regexes are quite small, so the memory overhead should be rather small too, as well as the compilation time.

Still, please make a final decision.